cache/lru/lru.go

```diff
@@ -0,0 +1,173 @@
+	size uint64
+
+	// ll is a doubly linked list which keeps track of recency of used elements
+	// by moving them to the front.
+	ll *list.List
+
+	// cache is a generic cache which allows us to find an elements position
+	// in the ll list from a given key.
+	cache elementMap
+
+	// mtx is used to make sure the Cache is thread-safe.
+	mtx sync.RWMutex
+}
+
+// NewCache return a cache with specified capacity, the cache's size can't
+// exceed that given capacity.
+func NewCache(capacity uint64) *Cache {
+	return &Cache{
+		capacity: capacity,
+		ll:       list.New(),
+		cache:    make(elementMap),
+	}
+}
+
+// evict will evict as many elements as necessary to make enough space for a new
+// filter to be inserted. sizeTotal is the amount of space evict needs to make.
+// It could end up evicting more than sizeTotal if the size of the least
+// recently used element to evict was bigger than the size needed to evict.
+func (c *Cache) evict(sizeTotal uint64) error {
+	if sizeTotal > c.capacity {
+		return fmt.Errorf(
+			"can't evict %v elements in size, since capacity is %v",
+			sizeTotal, c.capacity)
+	}
+
+	evicted := uint64(0)
+	for evicted < sizeTotal {
+		// We still need to evict some more elements.
+		if c.ll.Len() == 0 {
+			// We should never reach here.
+			return fmt.Errorf(
+				"all elements got evicted, yet still need to evict %v,"+
+					"likelihood of error during size calculation",
+				sizeTotal-evicted)
+		}
+
+		// Find the least recently used item.
+		if elr := c.ll.Back(); elr != nil {
+			// Determine lru item's size.
+			ce := elr.Value.(*entry)
+			es, err := ce.value.Size()
+			if err != nil {
+				return fmt.Errorf(
+					"couldn't determine size of existing cache value %v", err)
+			}
+
+			// Account for that element's removal in evicted and cache size.
+			evicted += es
+			c.size -= es
+
+			// Remove the element from the cache.
+			c.ll.Remove(elr)
+			delete(c.cache, ce.key)
+		}
+	}
+
+	return nil
+}
+
+// Put inserts a given (key,value) pair into the cache, if the key already
+// exists, it will replace value and update it to be most recent item in cache.
+func (c *Cache) Put(key interface{}, value cache.Value) error {
+	vs, err := value.Size()
+	if err != nil {
+		return fmt.Errorf("couldn't determine size of cache value: %v", err)
+	}
+	if vs > c.capacity {
+		return fmt.Errorf(
+			"can't insert entry of size %v into cache with capacity %v",
+			vs, c.capacity)
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	// If the filter already exists, remove it and decrease cache's size.
+	el, ok := c.cache[key]
+	if ok {
+		es, err := el.Value.(*entry).value.Size()
+		if err != nil {
+			return fmt.Errorf(
+				"couldn't determine size of existing cache value %v", err)
+		}
+		c.ll.Remove(el)
+		c.size -= es
+	}
+
+	// Then we need to make sure we have enough space for the filter, evict
+	// filters if we need more space.
+	if c.size+vs > c.capacity {
```

Comment:
`evict` should be change to take the size of the filter to insert. That way the `if` statement can be removed and this can be simplified to
```go
if err := c.evict(vs); err != nil {
```